internal/datacoord/import_checker.go

```diff
@@ -0,0 +1,394 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datacoord
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/samber/lo"
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/datacoord/broker"
+	"github.com/milvus-io/milvus/internal/proto/datapb"
+	"github.com/milvus-io/milvus/internal/proto/internalpb"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/metrics"
+	"github.com/milvus-io/milvus/pkg/util/tsoutil"
+)
+
+type ImportChecker interface {
+	Start()
+	Close()
+}
+
+type importChecker struct {
+	meta         *meta
+	broker       broker.Broker
+	cluster      Cluster
+	alloc        allocator
+	sm           Manager
+	imeta        ImportMeta
+	buildIndexCh chan UniqueID
+
+	closeOnce sync.Once
+	closeChan chan struct{}
+}
+
+func NewImportChecker(meta *meta,
+	broker broker.Broker,
+	cluster Cluster,
+	alloc allocator,
+	sm Manager,
+	imeta ImportMeta,
+	buildIndexCh chan UniqueID,
+) ImportChecker {
+	return &importChecker{
+		meta:         meta,
+		broker:       broker,
+		cluster:      cluster,
+		alloc:        alloc,
+		sm:           sm,
+		imeta:        imeta,
+		buildIndexCh: buildIndexCh,
+		closeChan:    make(chan struct{}),
+	}
+}
+
+func (c *importChecker) Start() {
+	log.Info("start import checker")
+	var (
+		stateTicker   = time.NewTicker(2 * time.Second)
+		collTicker    = time.NewTicker(30 * time.Second)
+		timeoutTicker = time.NewTicker(10 * time.Minute)
+	)
+	defer stateTicker.Stop()
+	defer collTicker.Stop()
+	defer timeoutTicker.Stop()
+	for {
+		select {
+		case <-c.closeChan:
+			log.Info("import checker exited")
+			return
+		case <-stateTicker.C:
+			jobs := c.imeta.GetJobBy()
+			for _, job := range jobs {
+				if job.GetSchema() == nil {
+					err := UpdateSchema(job, c.broker, c.imeta)
+					if err != nil {
+						continue
+					}
+				}
+				c.checkLackPreImport(job)
```

Comment:
How about add state in a job, such as PreImporting, Importing, etc.?
if job is Importing， you can skip checkLackPreImport